### PR TITLE
feat(belongs-to-many): allow creation of paranoid join tables

### LINKED ISSUE #11991: Paranoid as option when creating many-to-many association
## Issue Description
I want to be able to create a belongs-to-many association that is paranoid. The association should create a join table that has a deletedAt column. 

I found this in `lib/associations/belongs-to-many.js`. Why does it say that a paranoid join table doesn't make sense?
```js
    if (typeof this.through.model === 'string') {
      if (!this.sequelize.isDefined(this.through.model)) {
        this.through.model = this.sequelize.define(this.through.model, {}, Object.assign(this.options, {
          tableName: this.through.model,
          indexes: [], //we don't want indexes here (as referenced in #2416)
          paranoid: false, // A paranoid join table does not make sense
          validate: {} // Don't propagate model-level validations
        }));
      } else {
        this.through.model = this.sequelize.model(this.through.model);
      }
    }
```

### Describe the solution you'd like
I was thinking something along the lines of a `paranoid` key in the options object and you would create the association like: 
```js
Customer.belongsToMany(PhoneNumber, {
  through: 'CustomerPhoneNumbers',
  as: 'SecondaryPhoneNumbers',
  paranoid: true
})
```

### Describe alternatives/workarounds you've considered
A workaround is to define the join table as a separate model and have the model be paranoid. 
But this requires potentially many models that are only required because we want to add a deletedAt column.

### Is this issue dialect-specific?

- [X] No. This issue is relevant to Sequelize as a whole.
- [ ] Yes. This issue only applies to the following dialect(s): XXX, YYY, ZZZ

### Would you be willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time and I know how to start.
- [ ] Yes, I have the time but I don't know how to start, I would need guidance.
- [X] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

### PR DESCRIPTION
**Pull request by Leithal3**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md)?

### Description of change

add check of through options to allow creation of paranoid join tables
 - default is still false

no breaking changes

Closes #11991

### CODE REVIEW

**Comment by Leithal3** [comment]
Open to feedback. Happy to discuss reasoning.

Unsure how to fix the continuous-integration/appveyor/pr build - the log seems to indicate that a script file can't be found.

File: types/lib/associations/belongs-to-many.d.ts [hunk 1]
```diff
@@ -22,7 +22,13 @@ export interface ThroughOptions {
   /**
```
Comment (sushantdhiman) [inline types/lib/associations/belongs-to-many.d.ts:22]: Please add tests for these typing changes

Comment (Leithal3) [reply types/lib/associations/belongs-to-many.d.ts:22]: I have added a test, please let me know if the test I have added is inadequate or in the wrong place.

**Comment by sushantdhiman** [comment]
Please remove changes from lock file

**Pull request merged**